docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,256 @@
+* All autonomous interactions with the settings API must explicitly state the settings extension being interacted with, and at what version.
+* Migrations between settings representations as we know them today must cease to exist -- settings extensions must independently expose settings at supported interface versions and implement data migrations between them.
+* Settings extension interfaces, services, and configurations must all be represented in the build system, with the ability to express dependencies on these artifacts.
+
+## Settings Extensions
+This document provides an overview on the following:
+* Storage
+* Validation
+* Retrieval
+* Model migrations
+* System configuration via templating
+* Defaults generation
+* Extensions development with the Settings SDK
+
+### Setting Representation
+Today, Bottlerocket's `apiserver` utilizes models to perform strict validation of settings inputs, which are then serialized to JSON and stored in Bottlerocket's datastore.
+The models are defined using Rust structures and are statically compiled into the operating system, using Rust's type system and compiler to subsequently check interactions with these modelled objects.
+
+In principle, this will remain true -- Bottlerocket will continue to own settings objects which it stores as JSON, and changes will be validated against models on update.
+A primary difference here is that the settings objects will be opaque to Bottlerocket's core.
+Bottlerocket will understand that settings are JSON objects and be able to make modifications requested when, for example, calling `apiclient set`; however, the changes are then submitted to the settings extensions installed on the system in order to validate the proposed changes.
+
+### Extension Binaries
+All settings in a Bottlerocket host will be described by their own settings extension, which are defined by a binary packaged onto the system and symlinked into a common directory, as well as a configuration file placed in a corresponding `settings.d`-style directory.
+Extension binaries must respond to command line arguments following a specified protocol spoken by the Bottlerocket `apiserver`.
+The protocol is versioned, so all commands to an extension binary begin with the protocol version (in this document, all commands fall under the initial `proto1` protocol version -- so-named to avoid conflation with settings versions, which will be explained later).
+Command line arguments are used to signify the "request" being made of the extension.
+Exit codes will be used to signal the status of the request, with formatted output being delivered on `stdout`, and logs delivered on `stderr`.
+
+As an example, suppose a user wants to set the "message of the day" for their instance:
+```bash
+$ apiclient set motd="Hello, Bottlerocket!"
+```
+
+The apiserver would load the current value of `motd` and then create a new object based on the inputted string.
```

Comment:
You may be remembering an earlier form of this design where `get` requests would be routed to the extension to respond to. The idea here is that the API server can handle any arbitrary `get` request by referring to the stored JSON document.

Since the API server is also completely capable of handling arbitrary patches to JSON objects, it performs sets by changing the values requested by the user, and then just asking the settings extensions to validate them for correctness.